File: php/src/Models/RoutingRule.php
<?php

declare(strict_types=1);

namespace Rerout\Models;

use InvalidArgumentException;

/**
 * A Smart Link routing rule. When a rule matches the incoming request the link
 * resolves to the rule's `targetUrl` instead of the link's default destination.
 *
 * Used both in the {@see Link} response and as a request item on
 * {@see CreateLinkInput} / {@see UpdateLinkInput} (where it is rendered via
 * {@see RoutingRule::toArray()}).
 *
 * - `conditionType` is one of `country` / `device`.
 * - `conditionOp` is one of `is` / `is_not` / `in`.
 * - `conditionValue` is the value(s) to match against (e.g. `ZA`, `mobile`,
 *   or a comma-separated list for `in`).
 */
final readonly class RoutingRule
{
    /**
     * @param string $conditionType  Attribute to match — `country` or `device`.
     * @param string $conditionOp    Comparison operator — `is`, `is_not`, or `in`.
     * @param string $conditionValue Value(s) to match against.
     * @param string $targetUrl      Destination when the rule matches.
     */
    public function __construct(
        public string $conditionType,
        public string $conditionOp,
        public string $conditionValue,
        public string $targetUrl,
    ) {
    }

    /**
     * Parse a {@see RoutingRule} from an API JSON object.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        return new self(
            conditionType: self::requireString($data, 'condition_type'),
            conditionOp: self::requireString($data, 'condition_op'),
            conditionValue: self::requireString($data, 'condition_value'),
            targetUrl: self::requireString($data, 'target_url'),
        );
    }

    /**
     * Render to the JSON-serialisable shape the API expects.
     *
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'condition_type' => $this->conditionType,
            'condition_op' => $this->conditionOp,
            'condition_value' => $this->conditionValue,
            'target_url' => $this->targetUrl,
        ];
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function requireString(array $data, string $key): string
    {
        $value = $data[$key] ?? null;
        if (!is_string($value)) {
            throw new InvalidArgumentException("Expected string field '{$key}' in RoutingRule payload.");
        }

        return $value;
    }
}

File: dart/lib/src/models/routing_rule.dart
//
//  rerout
//  routing_rule.dart
//
//  Official Dart SDK for the Rerout branded-link API.
//
//  Created by Ngonidzashe Mangudya on 04/06/2026.
//  Copyright (c) 2026 Codecraft Solutions
//  Licensed under the MIT License
//  https://codecraftsolutions.co.za
//

import 'package:meta/meta.dart';

/// A geo/device routing rule (Smart Links).
///
/// When the rule's condition matches an incoming request, the redirect
/// resolves to [targetUrl] instead of the link's default destination.
///
/// - [conditionType] is `country` or `device`.
/// - [conditionOp] is `is`, `is_not`, or `in`.
@immutable
class RoutingRule {
  /// Creates a [RoutingRule].
  const RoutingRule({
    required this.conditionType,
    required this.conditionOp,
    required this.conditionValue,
    required this.targetUrl,
  });

  /// Parses a [RoutingRule] from the API JSON envelope.
  factory RoutingRule.fromJson(Map<String, dynamic> json) => RoutingRule(
    conditionType: json['condition_type'] as String,
    conditionOp: json['condition_op'] as String,
    conditionValue: json['condition_value'] as String,
    targetUrl: json['target_url'] as String,
  );

  /// What to match against — `country` or `device`.
  final String conditionType;

  /// Comparison operator — `is`, `is_not`, or `in`.
  final String conditionOp;

  /// Value(s) to compare against (e.g. `ZA`, `US,GB`, `mobile`).
  final String conditionValue;

  /// Destination when the rule matches.
  final String targetUrl;

  /// Converts this rule to the JSON map expected by the API.
  Map<String, dynamic> toJson() => {
    'condition_type': conditionType,
    'condition_op': conditionOp,
    'condition_value': conditionValue,
    'target_url': targetUrl,
  };

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is RoutingRule &&
          conditionType == other.conditionType &&
          conditionOp == other.conditionOp &&
          conditionValue == other.conditionValue &&
          targetUrl == other.targetUrl;

  @override
  int get hashCode =>
      Object.hash(conditionType, conditionOp, conditionValue, targetUrl);

  @override
  String toString() =>
      'RoutingRule($conditionType $conditionOp $conditionValue -> $targetUrl)';
}
